Fix generic_parameter_specialization_map in the presence of non-symmetric generic recursion [TG-7815]

NathanJPhillips opened the pull request:
This is a replacement for the closed PR #4913. It adopts a different approach as that approach was shown to be lacking.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- N/A My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

NathanJPhillips committed:
Change data storage for generic_parameter_specialization_mapt

This fixes the problem where we tried to unwind stacks in step that are not related, e.g. when a third generic class refers to two mutually recursive generic classes
Instead stacks are created per scope (generic context) and unwound individually
Also fixes problem where self-recursion isn't handled properly
Add test that would have triggered invariant in get_recursively_instantiated_type

NathanJPhillips committed:
Use type_try_dynamic_cast

NathanJPhillips committed:
Validate config is initialised when a member_offset_expr is constructed

This will save time hunting for the cause of problems when unit tests don't initialise the config.

NathanJPhillips committed:
Change naming of implicit type parameters for inner classes

Previously they duplicated the names for the type parameters of the outer classes

NathanJPhillips committed:
Give separate REQUIRE statements for each element in list

This will produce better error messages when tests fail.

smowton commented on jbmc/src/java_bytecode/select_pointer_type.cpp:
```diff
@@ -82,26 +81,25 @@ pointer_typet select_pointer_typet::specialize_generics(
-    }
+    const auto *array_element_type =
+      type_try_dynamic_cast<pointer_typet>(java_array_element_type(*subtype));
+    if(array_element_type == nullptr)
```
This commit ("Use type_try_dynamic_cast") does more than its message says (it also switches from `if(x) y` to `if(!x) return; y;` style)

smowton commented on jbmc/src/java_bytecode/Makefile:
```diff
@@ -5,6 +5,7 @@ SRC = assignments_from_json.cpp \
       ci_lazy_methods_needed.cpp \
       convert_java_nondet.cpp \
       expr2java.cpp \
+      generic_parameter_specialization_map.cpp \
```
Commit message for "Change data storage for generic_parameter_specialization_mapt" has got a bit out of hand -- the "Added ability to output..." bit probably doesn't need to be there if we're adding it at the same time as the map itself, and similarly "Add test..."

smowton commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.h:
```diff
@@ -0,0 +1,103 @@
+    map.map.container_to_specializations.size());
+  for(const auto &elt : map.map.param_to_container)
+  {
+    std::vector<irep_idt> &params = container_to_params[elt.second.first];
```
Using structs instead of `pair` would make it easier to understand what you're referring to here

smowton commented on jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.cpp:
```diff
@@ -151,17 +123,18 @@ void generic_parameter_specialization_map_keyst::insert_pairs_for_symbol(
   {
     return;
   }
+
```
Spacing fixups should be in their own commit

smowton commented on jbmc/src/java_bytecode/select_pointer_type.cpp:
```diff
@@ -39,47 +36,33 @@ pointer_typet select_pointer_typet::convert_pointer_type(
   if(parameter != nullptr)
   {
-    const irep_idt &parameter_name = parameter->get_name();
+    irep_idt parameter_name = parameter->get_name();
```
No need for this to change? `const irep_idt &` is pretty widely used for a const `irept`'s data, this probably isn't the place to fight that battle

NathanJPhillips replied:
I assign to it at the bottom of the loop.

smowton commented on jbmc/src/java_bytecode/generic_parameter_specialization_map_keys.h:
```diff
@@ -23,12 +23,10 @@ class generic_parameter_specialization_map_keyst
   /// Initialize a generic-parameter-specialization-map entry owner operating
   /// on a given map. Initially it does not own any map entry.
-  /// \param _generic_parameter_specialization_map: map to operate on.
+  /// \param generic_parameter_specialization_map: map to operate on.
```
Probably split commit "Code cleanup", e.g. "factor out get_final_name_component", one commit per variable rename, the stuff with `new_class_type`, the re-spacing are all pretty independent changes.

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -1200,6 +1200,17 @@ void mark_java_implicitly_generic_class_type(
     java_implicitly_generic_class_typet new_class_type(
       class_type, implicit_generic_type_parameters);
 
+    // Prepend existing parameters so choose those above any inherited
```
This commit should include a test

NathanJPhillips committed:
Remove duplicated brackets

NathanJPhillips committed:
Remove const modifier that doesn't do anything

NathanJPhillips committed:
Make explicit type parameters shadow implicit ones with the same names

Fixes problem where outer class type parameters take precedence over inner class ones
Previously field would be given the type of the argument assigned to A::T, not B::T.
class A<T> { class B<T> { T field; } }?

NathanJPhillips committed:
Remove prefix underscore

NathanJPhillips committed:
Extract get_final_name_component/get_without_final_name_component

NathanJPhillips committed:
Rename replacement_parameter_it

NathanJPhillips committed:
Invert the iterator check to reduce cognitive load caused by extended block scope

NathanJPhillips committed:
Use == instead of compare == 0

NathanJPhillips committed:
Update class type before assigning to class_symbol.type

Previously the code relied on the fact that class_type is a reference to class_symbol.type and so after a new type got copied assigned into class_symbol.type, class_type became a reference to the new value.

NathanJPhillips committed:
Use std::next to allow variable that isn't changed to be const

NathanJPhillips committed:
Indent TODO continuation lines

NathanJPhillips committed:
Use faster character version of rfind

NathanJPhillips committed:
Spacing

Add space between pre-checks and action
Remove space between abbreviations and their usage

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 833e1b70d70e4e75c6bdb9347c043c5d44dedf47).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120497981
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fcda9ce16368ab0dbf0fcfbe99e4c9872cdf9a54).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120498877
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

smowton commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp:
```diff
@@ -0,0 +1,76 @@
+        param_it->second.container_index == container_index,
+        "Not all type parameters assigned to same container");
+      INVARIANT(
+        param_it->second.param_index == param_index++,
```
Eeep! Let's not have `INVARIANT` conditions with side-effects, there are some compilation settings where the first arg is discarded

smowton commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp:
```diff
@@ -0,0 +1,76 @@
+    ", ",
+    [](const java_generic_parametert &param) { return param.get_name(); });
+  msg << ">";
+  std::string msg_str = msg.str();
```
Debugging code accidentally left in?

smowton commented on jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp:
```diff
@@ -187,7 +187,9 @@ SCENARIO(
     const class_typet &derived_class_type =
       require_type::require_complete_java_implicitly_generic_class(
-        derived_symbol.type, {"java::ContainsInnerClassGeneric::T"});
+        derived_symbol.type,
```
Double-check: do any of these tests exercise triple-nesting?
```
public class A<T> {
  class B {
    class C {
      T field;
    }
  }
}
```
That sort of thing?

NathanJPhillips replied:
The other test in that commit does: https://github.com/diffblue/cbmc/commit/aec9e89c82c9042b70342de9b9bedee96c5a99b4#diff-157ff943932394172d454093024a926cR120

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -1205,6 +1205,17 @@ void mark_java_implicitly_generic_class_type(
+    // Prepend existing parameters so choose those above any inherited
+    if(is_java_generic_class_type(class_type))
+    {
+      const java_generic_class_typet::generic_typest &class_type_params =
```
Commit message for "Make explicit type parameters shadow implicit ones with the same names" -- can you include a brief example of what you mean by precedence here? Are we talking `class A<T> { class B<T> { T field // which T is that? } }`?

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -1030,23 +1030,21 @@ static void find_and_replace_parameter(
-      replacement_parameter_full_name);
-  }
+  // A replacement parameter was found, update the identifier
+  const std::string &replacement_parameter_full_name =
```
Message: Sp congnative -> cognitive

smowton commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -1197,19 +1198,21 @@ void mark_java_implicitly_generic_class_type(
         base.type(), implicit_generic_type_parameters);
     }
+
+    class_symbol.type = new_class_type;
```
! This which was in the Code Cleanup commit suddenly looks like a bugfix -- the results of `find_and_replace_parameters` would previously have been wrongly discarded, right? Can we test the fix?

NathanJPhillips replied:
My assumption was that the thing we were changing was a reference to the same and this was just clarifying things. I'll check.

NathanJPhillips replied:
It was. I've updated the commit message to make this clear.

NathanJPhillips committed:
Fix indentation

smowton reviewed (approved):
Brilliant, thanks for putting so much effort into this one!

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f5f62cfb45573b1af6a7a15f77a44ccb69269b74).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/120739813
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

NathanJPhillips commented:
Passing CI for Diffblue bump here: https://github.com/diffblue/test-gen/runs/177289127

thomasspriggs commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.h:
```diff
@@ -0,0 +1,108 @@
+};
+
+template <typename ostream>
+ostream &operator<<(
```
:grey_question: Should `ostream` end in a `t` like `ostreamt` if it is a type?

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp:
```diff
@@ -0,0 +1,67 @@
+      const auto result = param_to_container.emplace(
+        parameter.get_name(), container_paramt{container_index, param_index++});
+      INVARIANT(
+        result.second, "Some but not all type parameters already mapped");
```
I'm not sure I understand the invariant message; there is no verb. :ox:

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp:
```diff
@@ -0,0 +1,67 @@
+      const auto param_it = param_to_container.find(parameter.get_name());
+      INVARIANT(
+        param_it != param_to_container.end(),
+        "Some but not all type parameters already mapped");
```
:ox:

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.cpp:
```diff
@@ -0,0 +1,67 @@
+        "Some but not all type parameters already mapped");
+      INVARIANT(
+        param_it->second.container_index == container_index,
+        "Not all type parameters assigned to same container");
```
:ox:

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.h:
```diff
@@ -0,0 +1,108 @@
+
+class generic_parameter_specialization_mapt
+{
+private:
```
:pick: the public section usually appears before the private one in cbmc

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.h:
```diff
@@ -0,0 +1,108 @@
+#include "expr2java.h"
+#include "java_types.h"
+
+class generic_parameter_specialization_mapt
```
This needs some documentation

romainbrenguier commented on jbmc/src/java_bytecode/generic_parameter_specialization_map.h:
```diff
@@ -0,0 +1,108 @@
+
+template <typename ostream>
+ostream &operator<<(
+  ostream &stm,
```
:question: Why does `ostream` need to be a template parameter instead of being just `std::ostream`?

NathanJPhillips replied:
It could be restricted to `std::ostream` but it doesn't need to be. I've added comment to that effect.

thomasspriggs commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -1142,8 +1142,9 @@ void mark_java_implicitly_generic_class_type(
 {
   const std::string qualified_class_name = "java::" + id2string(class_name);
   PRECONDITION(symbol_table.has_symbol(qualified_class_name));
+  // This will have its type changed
```
If you move the call to `get_writeable_ref` into your new `class_symbol.type = ...` line at the bottom, then this comment will no longer be necessary.

NathanJPhillips replied:
Then we would have two lookups though...

thomasspriggs reviewed (approved):
Looks like this fixes some stuff and I can't see any problems. Therefore this has my approval.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 186adb8aee29a624594d63fdf7ea824f7aa65753).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121391281
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.
